File: lee/src/main/java/org/example/lee/题目/栈/b_最小栈.java
package org.example.lee.题目.栈;

import java.util.Stack;

public class b_最小栈 {

    /**
     * min stack 压栈逻辑
     * https://leetcode.cn/problems/min-stack/solutions/674293/tu-li-zhan-shi-shuang-zhan-shi-xian-zui-fcwj5/?envType=study-plan-v2&envId=top-interview-150
     *
     * @author 刘家辉
     * @date 2024/01/17
     */
    class MinStack {
        private Stack<Integer> data;

        private Stack<Integer> help;

        public MinStack() {
            data = new Stack<>();
            help = new Stack<>();
        }

        public void push(int val) {
            data.push(val);
            if (help.isEmpty() || val <= help.peek()) {
                help.push(val);
            } else {
                help.push(help.peek());
            }

        }

        public void pop() {
            data.pop();
            help.pop();
        }

        public int top() {
            return data.peek();
        }

        public int getMin() {
            return help.peek();
        }
    }
}

File: lee/src/main/java/org/example/lee/题目/栈/b_最小栈.go
package 栈

// 常数时间内检索最小元素
type MinStack struct {
	main []int
	help []int
}

func Constructor() MinStack {
	return MinStack{
		main: []int{},
		help: []int{},
	}
}

func (this *MinStack) Push(val int) {
	this.main = append(this.main, val)
	if len(this.help) == 0 || this.help[len(this.help)-1] >= val {
		this.help = append(this.help, val)
	} else {
		this.help = append(this.help, this.help[len(this.help)-1])
	}
}

func (this *MinStack) Pop() {
	this.main = this.main[:len(this.main)-1]
	this.help = this.help[:len(this.help)-1]
}

func (this *MinStack) Top() int {
	return this.main[len(this.main)-1]
}

func (this *MinStack) GetMin() int {
	return this.help[len(this.help)-1]
}
